community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/internal/SessionStateMachine.java

```diff
@@ -889,7 +939,11 @@ private void before( Object attachment, Callback cb )
             cb.started( attachment );
         }
 
-        if( interruptCounter.get() > 0 )
+        if ( terminationMark.get() )
```

Comment:
Premature optimization, you honour! The terminationMark will be set extremely rarely, and therefore the lock will almost always be free to take, which is very fast. In particular grabbing the lock will be fast compared to downstream query execution. Apart from that, I vote for not nesting the checks as this introduces the theoretical possibility of escaping termination if the command that is processed during the setting of the terminationMark is a reset().